src/v/kafka/server/snc_quota_manager.cc

```diff
@@ -103,12 +105,33 @@ snc_quota_manager::snc_quota_manager()
         _shard_quota.eg.set_width(v);
     });
 
+    if (ss::this_shard_id() == quota_balancer_shard) {
+        _balancer_timer.set_callback([this] {
+            ssx::spawn_with_gate(_balancer_gate, [this] {
+                return quota_balancer_step().finally(
+                  [this] { maybe_arm_balancer_timer(); });
+            });
+        });
+    }
     vlog(klog.debug, "qm - Initial quota: {}", _shard_quota);
 }
 
-ss::future<> snc_quota_manager::start() { return ss::make_ready_future<>(); }
+ss::future<> snc_quota_manager::start() {
+    if (ss::this_shard_id() == quota_balancer_shard) {
+        _balancer_timer.arm(
+          ss::lowres_clock::now() + get_quota_balancer_node_period());
+    }
+    return ss::make_ready_future<>();
+}
 
-ss::future<> snc_quota_manager::stop() { return ss::make_ready_future<>(); }
+ss::future<> snc_quota_manager::stop() {
+    if (ss::this_shard_id() == quota_balancer_shard) {
+        _balancer_timer.cancel();
```

Comment:
I notice that there is no way to cancel in-progress work for shutdown, usually this is done via `ss::abort_source` and having your loop exit on condition that `request_abort` has been called.

In the worst case without this, the future returned by `stop()` will resolve in as long as it will take to execute one iteration of  `quota_balancer_step()`. Would you know how long this could be in the worst case? I see in previous comments you mention how in high load scenarios it could take a long time for a step to complete, in which case this would hold up shutdown.

Reply:
@graphcareful do you have a case in mind for which abort_source would speed up shutdown? it's nearly always the case that abort_source is useful in aborting things early like sleep(), semaphore::wait, condition_variable::wait, etc... but i don't recall seeing any of these usages. in the case of this PR we effectively have a simulated sleep with an ss::timer, but snc_quota_manager::stop will stop the timer and close the gate for an equivalent speedy shutdown.